File: Go/markov.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"strings"
)

const (
	nPrefix   = 2
	maxOutput = 200
	// Sentinel value that cannot organically appear in the state-table
	nonWord = "\n"
)

func main() {
	// The state-table that holds prefix phrases and their associated suffixes.
	st := make(map[string][]string)

	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)

	// Initialise the prefix slice with a non-word sentinel value to mark the
	// start of the document; useful during text generation.
	prefixes := make([]string, 0, nPrefix)
	for i := 0; i < nPrefix; i++ {
		prefixes = append(prefixes, nonWord)
	}

	// Build the state-table.
	for scanner.Scan() {
		suffix := scanner.Text()

		// The prefix key in the state-table is just the concatination of the
		// prefix strings.
		key := strings.Join(prefixes, "")
		st[key] = append(st[key], suffix)

		// Move the prefix window, and add the suffix.
		prefixes = append(prefixes[1:], suffix)
	}

	// Reset prefixes to sentinel values to indicate the beginning of the
	// document.
	prefixes = []string{}
	for i := 0; i < nPrefix; i++ {
		prefixes = append(prefixes, nonWord)
	}

	// Generate output
	for n := 0; n < maxOutput; n++ {
		key := strings.Join(prefixes, "")

		if suffixes, ok := st[key]; ok {
			suffix := suffixes[rand.Intn(len(suffixes))]
			fmt.Printf("%s ", suffix)

			prefixes = append(prefixes[1:], suffix)

		} else {
			break
		}
	}

	fmt.Printf("\n")
}

File: Python/markov.py
import random
import sys

def main():
    N_PREFIX = 2
    MAX_OUTPUT = 200
    NON_WORD = "\n"

    # This is the state-table. It associates prefix phrases with suffixes
    # they've appeared with in the input text.
    st = {}

    # Initialise the prefix window with a sentinal value that cannot naturally
    # occur in the state-table. This will help during text generation to handle
    # the beginning of the document correctly.
    prefixes = [NON_WORD] * N_PREFIX

    # Build the state-table.
    for line in sys.stdin:
        for suffix in line.split():
            key = "".join(prefixes)

            if key in st:
                st[key].append(suffix)
            else:
                st[key] = [suffix]

            # Move the prefix window and add the suffix.
            prefixes = prefixes[1:]
            prefixes.append(suffix)

    prefixes = [NON_WORD] * N_PREFIX

    for _ in range(MAX_OUTPUT):
        key = "".join(prefixes)

        if key in st:
            suffixList = st[key]
            suffix = random.choice(suffixList)

            prefixes = prefixes[1:]
            prefixes.append(suffix)

            print(suffix, end=" ")

        else:
            break

    print("\n")

if __name__ == "__main__":
    main()
